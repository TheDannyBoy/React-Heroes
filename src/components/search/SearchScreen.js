import React, { useMemo } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string'
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../selectors/getHero';

export const SearchScreen = ({ history }) => {
    // Get the hero keyword from queryParam
    const { search } = useLocation();
    const { q: query = '' } = queryString.parse(search);

    // Match the queryParam to the input value
    const [formValues, handleInputChange] = useForm({ searchText: query });
    const { searchText } = formValues;

    // Filter the heroes
    const heroesFiltered = useMemo(() => getHeroesByName(query), [query]);

    const handleSearch = (evt) => {
        evt.preventDefault();
        history.push(`?q=${searchText}`)
    };

    return (
        <>
            <div>SearchScreen</div>
            <hr />

            <div className='row'>
                <div className="col-12">
                    <h4>Search Form</h4>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input
                            type='text'
                            placeholder='Find your hero...'
                            className='form-control'
                            autoComplete='off'
                            name='searchText'
                            value={searchText}
                            onChange={handleInputChange}
                        />
                        <button
                            type='submit'
                            className='btn mt-1 btn-block btn-outline-primary'
                        >
                            Search
                        </button>
                    </form>
                </div>
            </div>
            <div className="row mt-5">
                <div className='col-12'>
                    <h4>Results</h4>
                    <hr />

                    {
                        !heroesFiltered.length
                            ?
                            <div className="alert alert-info">No heroes to show</div>
                            :
                            <div className='card-columns animate__animated animate__fadeIn'>
                                <ul>
                                    {
                                        heroesFiltered.map(hero =>
                                            <HeroCard key={hero.id} {...hero} />
                                        )
                                    }
                                </ul>
                            </div>
                    }
                </div>
            </div>
        </>
    );
}
