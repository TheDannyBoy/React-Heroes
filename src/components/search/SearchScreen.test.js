import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from './SearchScreen';

describe('SearchScreen', () => {
    test('Should show correctly with default valules', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('Should show Batman and the input value the same as the query string', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
    });

    test('Should show an error if there is no hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-info').text().trim()).toBe('No heroes to show');
    });

    test('Should show push history()', () => {
        const history = { push: jest.fn() };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path='/search' component={() => <SearchScreen history={history} />} />
            </MemoryRouter>
        );
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'hulk'
            }
        });
        wrapper.find('form').prop('onSubmit')({ preventDefault: jest.fn() });

        expect(history.push).toHaveBeenCalledWith('?q=hulk');
    });
});
