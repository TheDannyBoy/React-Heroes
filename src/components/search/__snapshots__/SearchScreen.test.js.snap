// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen Should show correctly with default valules 1`] = `
Array [
  <div>
    SearchScreen
  </div>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero..."
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </div>,
  <div
    className="row mt-5"
  >
    <div
      className="col-12"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        No heroes to show
      </div>
    </div>
  </div>,
]
`;
